sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/QueryPlanConstraints.scala

```diff
@@ -77,20 +77,21 @@ trait QueryPlanConstraints { self: LogicalPlan =>
     constraint match {
       // When the root is IsNotNull, we can push IsNotNull through the child null intolerant
       // expressions
-      case IsNotNull(expr) => scanNullIntolerantAttribute(expr).map(IsNotNull(_))
+      case IsNotNull(expr) => scanNullIntolerantField(expr).map(IsNotNull(_))
       // Constraints always return true for all the inputs. That means, null will never be returned.
       // Thus, we can infer `IsNotNull(constraint)`, and also push IsNotNull through the child
       // null intolerant expressions.
-      case _ => scanNullIntolerantAttribute(constraint).map(IsNotNull(_))
+      case _ => scanNullIntolerantField(constraint).map(IsNotNull(_))
```

Comment:
Previously IsNotNull constraints are Attribute-specific, why do we need to expand it to `ExtractValue`?

Reply:
I think the best way to explain is through an example. Suppose we have a Parquet file with two string columns: `name.first` and `name.last`. We'll register this file as a temp view with name `people`. Now consider the following query:

```sql
select name.first from people where name.first = 'Michael'
```

The only column we need to read from the file to answer this query is `name.first`.

Without the change to `QueryPlanConstraints.scala`, the optimized plan will look like:

```
Project [name#19.first AS first#30]
+- Filter (isnotnull(name#19) && (name#19.first = Michael))
   +- Relation[name#19] parquet
```

Evaluating `isnotnull(name#19)` requires the entire name struct. So both `name.first` and `name.last` will be read.

With the change, the optimized pln will look like:

```
Project [name#19.first AS first#30]
+- Filter (isnotnull(name#19.first) && (name#19.first = Michael))
   +- Relation[name#19] parquet
```

Now the `isnotnull` expression requires only the `name.first` column. The `name.last` column will not be read.

I don't believe this is covered in the current unit tests, so I will add or modify a test to cover it.

Note that in either case the `LogicalRelation` is shown as `Relation[name#19] parquet`, which is confusing. The string output for a logical relation returns attributes, not fields. Perhaps we should do something about that? So in the latter case it would print `Relation[name#19.first] parquet` instead.

Reply:
Incidentally, I have noticed that I need to update the code comments where I made the changes. I will push a commit with revised comments.

Reply:
> Incidentally, I have noticed that I need to update the code comments where I made the changes. I will push a commit with revised comments.

I've pushed the commit with the revised documentation: https://github.com/apache/spark/pull/16578/commits/38cec5f05066f75d868b64d360b04edb18dcebeb.

Reply:
> I don't believe this is covered in the current unit tests, so I will add or modify a test to cover it.

I added the following test to cover this case:

https://github.com/apache/spark/blob/88786d3eaf9d3a3d2c80c2235c5014074ade3dc1/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala#L62-L79